rust/worker/src/segment/types.rs

```diff
@@ -864,21 +867,147 @@ pub async fn materialize_logs(
 // This needs to be public for testing
 #[allow(async_fn_in_trait)]
 pub trait SegmentWriter {
+    type Flusher: SegmentFlusher + Sync + Send;
+
+    fn get_id(&self) -> SegmentUuid;
     fn get_name(&self) -> &'static str;
     fn apply_materialized_log_chunk(
         &self,
         record_segment_reader: &Option<RecordSegmentReader>,
         materialized_chunk: &MaterializeLogsResult,
     ) -> impl Future<Output = Result<(), ApplyMaterializedLogError>> + Send;
-    async fn commit(self) -> Result<impl SegmentFlusher, Box<dyn ChromaError>>;
+    fn finish(&mut self) -> impl Future<Output = Result<(), Box<dyn ChromaError>>> + Send {
+        async { Ok(()) }
+    }
+    fn commit(self) -> impl Future<Output = Result<Self::Flusher, Box<dyn ChromaError>>> + Send;
 }
 
 // This needs to be public for testing
 #[async_trait]
 pub trait SegmentFlusher {
+    fn get_id(&self) -> SegmentUuid;
+    fn get_name(&self) -> &'static str;
     async fn flush(self) -> Result<HashMap<String, Vec<String>>, Box<dyn ChromaError>>;
 }
 
+#[derive(Clone, Debug)]
+pub enum ChromaSegmentWriter<'bf> {
```

Comment:
Why haven enum and a trait? This feels really odd to me.

Reply:
thanks for the call out
enum dispatch seems to generally be much nicer to work with since it's a sized type, especially when dealing with futures
I figured I would keep the existing trait to force consistent implementations across the different segment writer types, but I think you're right that it's unnecessary now. I'll remove the trait.